### ISSUE TITLE
Suggestion: Get Physical Cores (Not Just Logical)

### ISSUE THREAD

**Issue by time-killer-games** [libsdl-org/SDL#8116]
Basically we have this, which can only get the logical core count:
```
SDL_GetCPUCount()
```
But it would be better to support physical core count as well, so:
```
SDL_GetCPULogicalCount()
SDL_GetCPUPhysicalCount()
```
I wrote code which does this, and I would like to create a pull request, though I am not sure if how I am doing it would be acceptable for your coding practice standards, as for a lot of platforms I am reading output from shell scripts / subprocesses: https://github.com/time-killer-games/ween/blob/153eb027eeaadb468222bac2c822faf53302dbed/system.cpp#L1322-L1449 Plus, on NetBSD / OpenBSD there is no official means to do this, so you would need inline assembly. This would imply x86 support only, though it could be managed with other architectures, it would likely not be worth it to write code for every single exotic architecture which are supported by NetBSD and OpenBSD. 

For those platforms I would think it would be safe to assume the same number of cores for both physical and logical, and perhaps throw a warning to output letting the user know this value being used might not be all that accurate in some obscure cases, as there are some ARM boards with this issue. That, and there is also the fact my code is currently written in C++, so it will need to be adapted to be written completely in C.

**Comment by slouken** [comment]
We definitely don't want to be calling out to external programs for this.

@icculus, thoughts?

**Comment by icculus** [comment]
Ugh, yeah, we don't want to parse dmesg output, for sure.

**Comment by time-killer-games** [comment]
Isn't it true reading dmesg output sometimes can be prevented from the operating system via some type of configuration? All I know is the platform I am doing that on (DragonFly BSD) doesn't disable this out of the box, at least.

**Comment by time-killer-games** [comment]
I was thinking if I open a pull request I could at least do this for platforms which don't rely on running programs like my current code does on certain platforms, though I'm not so sure I'd have the patience to rewrite the platform implementations which I do have relying on that, so those platforms would be up to someone else ideally to implement.

**Comment by slouken** [comment]
> Isn't it true reading dmesg output sometimes can be prevented from the operating system via some type of configuration? All I know is the platform I am doing that on (DragonFly BSD) doesn't disable this out of the box, at least.

Yes, but a larger problem is that the dmesg buffer can wrap and the boot messages you're looking for can be overwritten and not available.

**Comment by slouken** [comment]
> Alternatively, we could just support exclusively x86 platforms via inline assembly on non-macOS platforms and also support macOS (both x86 and ARM) via sysctlbyname() and nothing else (perhaps as a starting goal?). I already have code for all that and we wouldn't need to write anything new, as none of this uses execution of programs.

That seems reasonable to me.

**Comment by time-killer-games** [comment]
You guys are welcome to use my code and remove copyright mentions of myself and change the license as needed, this is what I'd do in my pull request, but if I put off making a pr too long for your liking, feel free to make the changes yourself.

My depression has been preventing me from functioning in all aspects of life, so I don't expect doing it soon.

**Comment by Kontrabant** [comment]
For an arch agnostic way of getting the CPU topology, Linux exposes it via `/sys/devices/system/cpu/*`, and on Windows you can use `GetLogicalProcessorInformation()` (you need to go into the registry for performance info though).

**Comment by time-killer-games** [comment]
> For an arch agnostic way of getting the CPU topology, Linux exposes it via `/sys/devices/system/cpu/*`, and on Windows you can use `GetLogicalProcessorInformation()` (you need to go into the registry for performance info though).

The only problem with that is it covers only two of the many platforms SDL supports, so we would effectively need to find a way to do the same for all other potential platforms involved, which could be a bit of a challenge.

**Comment by slime73** [comment]
Would it make sense to differentiate between big and little core counts as well, since that sort of separation is becoming more popular even on laptop and desktop CPUs? Or would that be getting a bit too much into the weeds maybe?

I'm kind of curious what the use case is for non-HT core count, without more detailed info and control like core pinning.

**Comment by time-killer-games** [comment]
> Would it make sense to differentiate between big and little core counts as well, since that sort of separation is becoming more popular even on laptop and desktop CPUs? Or would that be getting a bit too much into the weeds maybe?
> 
> I'm kind of curious what the use case is for non-HT core count, without more detailed info and control like core pinning.

I think such information wouldn't be as useful as getting physical core count, though, I'm not in charge here, and only contributed to SDL like maybe 3 lines of code to OpenBSD, so not a good voice on the matter. If anyone of more authority likes your idea, and it picks up attention, who knows, someone might add it to SDL3. :)

**Comment by slime73** [comment]
I guess I'm mostly curious what people would use physical core count for, as someone who might expose this same API in my framework. Most (all?) uses I can think of are pretty specialized and dependent on hardware so extra context like big/little counts could be helpful there, but maybe I'm missing some big use cases.

**Comment by time-killer-games** [comment]
> I guess I'm mostly curious what people would use physical core count for, as someone who might expose this same API in my framework. Most (all?) uses I can think of are pretty specialized and dependent on hardware so extra context like big/little counts could be helpful there, but maybe I'm missing some big use cases.

To be honest, I'm no expert on processors and cores. I initially wrote my code to get logical and physical core count by request of a client who needed it for their work. I don't actually know what aspect of getting physical core count is actually considered useful, much less what the use case would be, as again, I'm not a good source of knowledge for this. I saw a feature SDL had (get logical count) and a feature it lacked (physical count) and figured I'd file a suggestion now that SDL3 is here and there will be API changes.

As far as big and little core counts, I figured it was less important because I saw less OS-level support for getting that info than even getting physical cores, though I have faulty memory and also did very limited research on it thus far.
